### PR TITLE
refactor: link components

### PR DESCRIPTION
**Pull request by nataliepina**
Converting link components to functional components with updated types and additional cleanup.

Refactoring the components to move away from class components and better prop types to prepare for React 18.

Refactored components include:
- FocusStyleManager
- UnstyledLink
- Link
- UIKItProvider
- UIKitThemeProvider
- customLink

Some of these components impact each other which is why they are grouped together in these changes.

## Testing

Regression testing on the stories for the components listed above. 

Button components should have the same functionality, they should still have an outline or underline with keyboard navigation. 

## Checklist for PR creator

- [ ] If any new components were added, there are exported from `packages/index.ts`
- [ ] If this PR is associated with a JIRA, it is mentioned in commit message footer ("Closes …")
- [ ] If this PR contains breaking changes, is stated in commit message body ("BREAKING CHANGE: …")
- [x] Info for applicable sections above is provided

### CODE REVIEW

File: packages/link/components/UnstyledLink.tsx [hunk 1]
```diff
@@ -1,32 +1,23 @@
 import React from "react";
 import { ExpandedLinkProps } from "../types";
-import { LinkComponentContext } from "../../uiKitProvider/link/context";
 
-class UnstyledLink extends React.Component<ExpandedLinkProps> {
-  static contextType = LinkComponentContext;
-  render() {
-    const LinkComponent = this.context;
+const UnstyledLink = (props: ExpandedLinkProps) => {
+  const { href, url, target, openInNewTab, context, children, ...rest } = props;
```
Comment (vacas5) [inline packages/link/components/UnstyledLink.tsx:5]: context won't be a prop here.

File: packages/link/components/UnstyledLink.tsx [hunk 2]
```diff
@@ -1,32 +1,23 @@
...
 
-    if (LinkComponent) {
-      return <LinkComponent {...this.props} />;
-    }
+  // Context is used for `UIKitProvider` link delegation
+  const LinkComponent = context;
```
Comment (vacas5) [inline packages/link/components/UnstyledLink.tsx:8]: try `const LinkComponent = React.useContext(LinkComponentContext)` here and remove context from props.

File: packages/link/types.ts [hunk 3]
```diff
@@ -2,13 +2,16 @@ import React from "react";
...
    */
   url?: string;
   /**
-   * if the `url` prop is set, this will open that link in a new tab
+   * If the `url` prop is set, this will open that link in a new tab.
    */
   openInNewTab?: boolean;
+  children?: React.ReactNode | React.ReactNode[];
+  className?: string | undefined;
+  context?: LinkComponent | undefined;
```
Comment (vacas5) [inline packages/link/types.ts:14]: you'll want to remove this to maintain the existing functionality

**Review by vacas5 — CHANGES_REQUESTED** [review]
Not exactly how you want to refactor the context from the class component

File: packages/link/components/UnstyledLink.tsx (hunk 2 above)
Comment (nataliepina) [reply packages/link/components/UnstyledLink.tsx:8]: Thanks for taking a look at this. Looks like this does the trick here 👍🏻

**Comment by nataliepina** [comment]
Screen Shot 2022-09-12 at 2 03 15 PM

I've made some additional updates to try to fix the issues with the link delegate story under the `UIKitProvider`. 
Unless there is something else to address the undefined 'className'.
I have been following the trail of errors. It looks like the next thing to update and convert might be `interactionMode.ts`, may be able to turn this into a hook. 
The issue is around the FocusStyleManager set on the Button component which the link delegate uses. I suspect it might have something to do with this being a class and the use of `this` and overall class-based logic.

**Review by vacas5 — APPROVED** [review]
looks like some other things changed with this PR too. I'll need a bit in the morning to look at it, but the other changes look good

**Comment by nataliepina** [comment]
@vacas5 
Sadly the UIKitProvider / Link Delegate story has the error above still.

Still WIP because of this, but if anyone wants to poke around and try to fix it they are very much welcome to!

**Review by vacas5 — APPROVED** [review]

**Pull request merged**